tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -6421,8 +6466,9 @@ bool HLSLExternalSource::MatchArguments(
       }
       pNewType = objectElement;
     }
-    else if (pArgument->uLegalComponentTypes == LICOMPTYPE_TEXTURE2D
-      || pArgument->uLegalComponentTypes == LICOMPTYPE_TEXTURE2DARRAY) {
+    else if (i != 0 && Template[pArgument->uTemplateId] == AR_TOBJ_OBJECT) {
+      // For object parameters, just use the argument type
+      // Return type is assigned below
```

Comment:
I don't think this code will correctly initialize the object type for second matching argument.  Instead, it's likely to just use the input argument type, which will match the input argument of course, but won't be able to detect when two arguments don't exactly match when they are the same `ArBasicKind`.  I made a related comment above regarding the change in CombineBasicTypes.

In order to match, I think we have to set the type to the type of the target argument at index `pArgument->uTemplateId` when this the target argument isn't this one.  Also, assert that the target argument is always `<=` this one (which it is in this case).